gradoop-temporal/src/main/java/org/gradoop/temporal/model/impl/operators/metric/functions/GroupDegreeTreesToAggregateDegrees.java

```diff
@@ -0,0 +1,122 @@
+   *
+   * @param aggregateType the aggregate type to use (min,max,avg).
+   */
+  public GroupDegreeTreesToAggregateDegrees(AggregateType aggregateType) {
+    this.aggregateType = aggregateType;
+  }
+
+  @Override
+  public void reduce(Iterable<Tuple2<GradoopId, TreeMap<Long, Integer>>> iterable,
+                     Collector<Tuple2<Long, Integer>> collector) throws Exception {
+
+    // init necessary maps and set
+    HashMap<GradoopId, TreeMap<Long, Integer>> degreeTrees = new HashMap<>();
+    HashMap<GradoopId, Integer> vertexDegrees = new HashMap<>();
+    SortedSet<Long> timePoints = new TreeSet<>();
+
+    // convert the iterables to a hashmap and remember all possible timestamps
+    for (Tuple2<GradoopId, TreeMap<Long, Integer>> tuple : iterable) {
+      degreeTrees.put(tuple.f0, tuple.f1);
+      timePoints.addAll(tuple.f1.keySet());
+    }
+
+    int numberOfVertices = degreeTrees.size();
+
+    // Add default times
+    timePoints.add(Long.MIN_VALUE);
+
+    for (Long timePoint : timePoints) {
+      // skip last default time
+      if (Long.MAX_VALUE == timePoint) {
+        continue;
+      }
+      // Iterate over all vertices
+      for (Map.Entry<GradoopId, TreeMap<Long, Integer>> entry : degreeTrees.entrySet()) {
+        // Make sure the vertex is registered in the current vertexDegrees capture
+        if (!vertexDegrees.containsKey(entry.getKey())) {
+          vertexDegrees.put(entry.getKey(), 0);
+        }
+
+        // Check if timestamp is in tree, if not, take the lower key
+        if (entry.getValue().containsKey(timePoint)) {
+          vertexDegrees.put(entry.getKey(), entry.getValue().get(timePoint));
+        } else {
+          Long lowerKey = entry.getValue().lowerKey(timePoint);
+          if (lowerKey != null) {
+            vertexDegrees.put(entry.getKey(), entry.getValue().get(lowerKey));
+          }
+        }
+      }
+
+      // Here, every tree with this time point is iterated. Now we need to aggregate for the current time.
+      Optional<Integer> opt;
+      Optional<Integer> opt2;
+      switch (aggregateType) {
+      case MIN:
+        opt = vertexDegrees.values().stream().reduce(Math::min);
+        opt.ifPresent(integer -> collector.collect(new Tuple2<>(timePoint, integer)));
+        break;
+      case MAX:
+        opt = vertexDegrees.values().stream().reduce(Math::max);
+        opt.ifPresent(integer -> collector.collect(new Tuple2<>(timePoint, integer)));
+        break;
+      case AVG:
+        opt = vertexDegrees.values().stream().reduce(Math::addExact);
+        opt.ifPresent(integer -> collector.collect(
+              new Tuple2<>(timePoint, (int) Math.ceil((double) integer / (double) numberOfVertices))));
+        break;
+      case RANGE:
+        opt = vertexDegrees.values().stream().reduce(Math::max);
+        opt2 = vertexDegrees.values().stream().reduce(Math::min);
+        opt.flatMap(max -> opt2.map(min -> max - min));
```

Comment:
This is not working. It does not change the value of the opt. Since the expected values in the UnitTests are wrong for BOTH, t4, you were possibly not aware of this.

Reply:
```
int maxDegree = vertexDegrees.values().stream().reduce(Math::max).orElse(0);
int minDegree = vertexDegrees.values().stream().reduce(Math::min).orElse(0);
collector.collect(new Tuple2<>(timePoint, maxDegree - minDegree));
```